File: 백준/Gold/1720. 타일 코드/타일 코드.py
# 타일 코드
import sys
input = sys.stdin.readline

n = int(input())
dp = [1] * 31 # 완전 대칭인 타일의 갯수
dp[2] = 3

dpOrigin = [0] * 31 # 모든 타일의 갯수
dpOrigin[1] = 1
dpOrigin[2] = 3


for i in range(3, n+1):
    dpOrigin[i] = dpOrigin[i-1] + dpOrigin[i-2] * 2

for i in range(4, n+1):
    dp[i] = dp[i-2] + dp[i-4] * 2


print(dpOrigin[n] - (dpOrigin[n] - dp[n]) // 2)

File: 백준/Gold/1720. 타일 코드/타일 코드.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws  Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] dp = new int[31];
        Arrays.fill(dp, 1);
        dp[2] = 3;

        int[] dpOrigin = new int[31];
        Arrays.fill(dpOrigin, 1);
        dpOrigin[2] = 3;

        for (int i=3; i <= n; i++){
            dpOrigin[i] = dpOrigin[i-1] + dpOrigin[i-2]*2;
        }

        for (int i=4; i <= n; i++) {
            dp[i] = dp[i-2] + dp[i-4]*2;
        }

        System.out.println(dpOrigin[n] - (dpOrigin[n] - dp[n]) / 2);
    }
}
